crossplane-contrib/terrajet
# Reproduce state in every reconciliation using the managed resource fields

### Description of your changes

With this change, we are able to reproduce tf state in every reconcile. Mechanism is roughly as the following:
* Dump parameters, observation and ID into state.
  * We dump everything we have because `terraform import` does not cover all information, like the non-sensitive values that are available only during creation.
* Run `terraform refresh` to get the current state of the world, which will update only the things that don't match.
  * At this point, we have the current state as tfstate file and desired state in our CR parameters.

We had to store [`private` attribute](https://serverfault.com/questions/981661/what-is-the-meaning-of-private-attribute-in-tfstate-for-a-dynamodb-instance) of the state object (usually a simple json with schema version info) in annotations because it's arbitrary metadata information that provider uses and cannot be regenerated since it's neither parameter nor observation.

While I'm working on this issue, I tried not to touch directly unrelated things but filled out https://github.com/crossplane-contrib/terrajet/issues/36 for those things and opened small other PRs for smaller refactorings.

Fixes https://github.com/crossplane-contrib/terrajet/issues/12

I have:


### How has this code been tested

Creation and stable state:
* Create the resource.
* Delete workspace folder
* Check whether it's recreated and everything is working.

Update:
* Create the resource.
* Delete workspace folder.
* Update CR instance.
* Check whether workspace is reproduced and update is issued.
